examples/hitl/rearrange_v2/object_state_manipulator.py

```diff
@@ -122,36 +128,41 @@ def get_action(
 
     def can_execute_action(
         self, state_name: str, state_value: Any, object_handle: str
-    ) -> Tuple[bool, str]:
+    ) -> ObjectStateChangeResult:
         """
         Test whether an action can be executed on the specified object.
         Returns whether the action can be executed, and an error message if it can't.
         """
         action = self.get_action(state_name, state_value)
         if action is not None:
-            return action.can_modify_state_impl(
+            result = action.can_modify_state_impl(
                 self._sim,
                 self._agent_index,
                 object_handle,
                 state_name,
                 self._maximum_distance,
                 self._world._metadata_interface,
             )
-        return (False, f"Undefined action: '{state_name} -> {state_value}'.")
+            return ObjectStateChangeResult(
+                result.succeeded, result.error_message_user
+            )
+        return ObjectStateChangeResult(
+            False, f"Undefined action: '{state_name} -> {state_value}'."
+        )
 
     def try_execute_action(
         self, state_name: str, state_value: Any, object_handle: str
-    ) -> Tuple[bool, str]:
+    ) -> ObjectStateChangeResult:
         """
         Try executing an action on the specified object.
         Returns whether the action was successful, and an error message if it wasn't.
         """
-        success, error_message = self.can_execute_action(
+        result = self.can_execute_action(
             state_name, state_value, object_handle
         )
-        if success:
+        if result.success:
             self.set_object_state(object_handle, state_name, state_value)
-        return (success, error_message)
+        return ObjectStateChangeResult(result.success, result.error_message)
```

Comment:
If result is not a success, is the error communicated anywhere to the user?

Reply:
No, in this case, it's just failing silently.

The UI calls `can_execute_action` every frame to check whether the user can execute the action. The message is passed here to be reactive to user's actions.